includes/Core/Util/Migration_1_8_1.php

```diff
@@ -0,0 +1,257 @@
+		// steps to resolve will be shown.
+		$credentials  = $this->authentication->credentials()->get();
+		$google_proxy = new Google_Proxy( $this->context );
+		$response     = wp_remote_post(
+			$google_proxy->url( '/notifications/mark/' ),
+			array(
+				'body' => array(
+					'site_id'            => $credentials['oauth2_client_id'],
+					'site_secret'        => $credentials['oauth2_client_secret'],
+					'notification_id'    => 'verification_leak',
+					'notification_state' => 'required',
+					// This is a special parameter only supported for this
+					// particular notification.
+					'identifiers'        => implode( ',', $unauthorized_identifiers ),
+				),
+			)
+		);
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$response_code = wp_remote_retrieve_response_code( $response );
+		if ( 200 !== $response_code ) {
+			$body    = wp_remote_retrieve_body( $response );
+			$decoded = json_decode( $body, true );
+			return new WP_Error( $response_code, ! empty( $decoded['error'] ) ? $decoded['error'] : $body );
+		}
+		return true;
+	}
+
+	/**
+	 * Checks for any users that are verified without proper authorization and
+	 * clears all their Site Kit data.
+	 *
+	 * @since 1.8.1
+	 *
+	 * @return array List of email addresses for the unauthorized users.
+	 */
+	private function clear_unauthorized_verified_users() {
+		global $wpdb;
+
+		$unauthorized_identifiers = array();
+		$profile                  = new Profile( $this->user_options );
+
+		// Store original user ID to switch back later.
+		$backup_user_id = $this->user_options->get_user_id();
+
+		// Iterate through all users verified via Site Kit.
+		foreach ( $this->get_verified_users() as $user ) {
+			$user_id = (int) $user->ID;
+			$this->user_options->switch_user( $user_id );
+
+			// If the user has setup access, there is no problem.
+			if ( user_can( $user_id, Permissions::SETUP ) ) {
+				continue;
+			}
+
+			// Try to get profile email, otherwise fall back to WP email.
+			if ( $this->authentication->profile()->has() ) {
+				$unauthorized_identifiers[] = $this->authentication->profile()->get()['email'];
+			} else {
+				$unauthorized_identifiers[] = $user->user_email;
+			}
+
+			$prefix = $this->user_options->get_meta_key( 'googlesitekit\_%' );
+			// phpcs:ignore WordPress.DB.DirectDatabaseQuery
+			$wpdb->query(
+				$wpdb->prepare( "DELETE FROM $wpdb->usermeta WHERE user_id = %d AND meta_key LIKE %s", $user_id, $prefix )
+			);
+			wp_cache_delete( $user_id, 'user_meta' );
+		}
+
+		// Restore original user ID.
+		$this->user_options->switch_user( $backup_user_id );
+
+		return $unauthorized_identifiers;
+	}
+
+	/**
+	 * Gets all users that are verified via Site Kit.
+	 *
+	 * @since 1.8.1
+	 *
+	 * @return WP_User[] User objects of verified users. Maximum of 20.
+	 */
+	private function get_verified_users() {
+		return get_users(
+			array(
+				'number'     => 20,
+				'meta_query' => array( // phpcs:ignore WordPress.DB.SlowDBQuery
```

Comment:
Since we're specifically looking for users who shouldn't be verified, should we also add `role__not_in` with `array( 'administrator' )` so we don't potentially hit the limit with users we know should be capable?

Reply:
Using the role wouldn't be reliable, we should rely on capability checks. If a developer intentionally modified the site to also enable the permission for e.g. editors, we should cater for that. The only way to properly deal with capabilities like that is on the code level, these permission checks are made in `clear_unauthorized_verified_users()`.